Add support for creating a CSPC template

vharsh opened the pull request:
# Summary

For a k8s cluster which has CStor installed, this CLI takes some arguments and generates a Pool-Cluster YAML to help user create a CStor pool of a certain kind via flags

fixes: #58 

```bash
$ kubectl openebs generate cspc --nodes=shubham
apiVersion: cstor.openebs.io/v1
kind: CStorPoolCluster
metadata:
  creationTimestamp: null
  generateName: cstor
  namespace: openebs
spec:
  pools:
  - dataRaidGroups:
    - blockDevices:
      # /var/openebs/sparse/0-ndm-sparse.img  10GB
      - blockDeviceName: sparse-6b277da87b7487e501c03ea0001d6d92
    nodeSelector:
      kubernetes.io/hostname: shubham
    poolConfig:
      dataRaidGroupType: stripe
```

## Pending items
- [ ] Unit testing
- [x] Implementing `makePools` for other poolTypes
- [ ] Adding capacity restriction for the pools

## Supported poolTypes

- [x] stripe
- [x] mirror
- [ ] raidz
- [ ] raidz2

Commit:
Add starter code for template generation

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

Commit:
Add disk generation

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

shubham14bajpai committed:
Add yaml string parsing

Signed-off-by: shubham <shubham14bajpai@gmail.com>

Commit:
Add support for poolType=mirror

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

Commit:
Add flaky unit tests for mirrored poolType

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

Commit:
Use utility func to humanize sizes

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

vharsh commented on pkg/client/k8s.go:
```diff
@@ -493,3 +481,44 @@ func (k K8sClient) GetBatchJobs() (*batchV1.JobList, error) {
+
+// GetDeploymentList returns the deployment-list with a specific
+// label selector query
+func (k K8sClient) GetDeploymentList(labelSelector string) (*appsv1.DeploymentList, error) {
```
This function got added via rebase

vharsh commented on pkg/generate/cspc.go:
```diff
@@ -0,0 +1,215 @@
+
+// isPoolValid checks if a CStor pool is valid
+func isPoolTypeValid(raid string) bool {
+	if raid == "stripe" || raid == "mirror" || raid == "raidz" || raid == "raidz2" {
```
please clarify if `mirror` or `mirrored` both are accepted terms and can be used interchangeably

mittachaitu replied:
The supported type is "mirror"... refer [apis](https://github.com/openebs/api/blob/ecd9a8a61d3ef0b2e3cc7a9c2d304eea9b5ce6d2/pkg/apis/cstor/v1/cstorpoolcluster.go#L25-L37) if required any clarification.

vharsh commented on pkg/generate/cspc.go:
```diff
@@ -0,0 +1,215 @@
+}
+
+// CSPC takes eligible nodes, number of devices and poolType to create a pool cluster template
+func CSPC(c *client.K8sClient, nodes []string, devs int, poolType string) (*cstorv1.CStorPoolCluster, string, error) {
```
I've added the structure, YAML string, error as return values to help while writing tests, the structure isn't used by callers, but it can be used later to augment some functionality later

vharsh commented on pkg/generate/cspc.go:
```diff
@@ -0,0 +1,215 @@
+	// assume CSTOR's OPENEBS_NAMESPACE has all the relevant blockdevices
+	c.Ns = cstorNS
+	if err != nil {
+		return nil, "", fmt.Errorf("cannot find an active cstor installation")
```
Please see if this error message is appropriate, my idea is, if the CSI driver labeled pod is missing and is not in a Running state, there's no active cstor installation.

vharsh commented on pkg/generate/cspc.go:
```diff
@@ -0,0 +1,215 @@
+		return nil, "", fmt.Errorf("cannot find an active cstor installation")
+	}
+	// 1. Validate nodes & poolType, fetch disks
+	nodeList, err := c.GetNodes(nodes, "", "")
```
Should only `status.Phase=Running` nodes be listed?

vharsh replied:
TODO: @vharsh, evaluate the cons of filtering out bad nodes with respect to code, UT & product usage.

vharsh commented on pkg/generate/cspc_test.go:
```diff
@@ -0,0 +1,143 @@
+
+
+
+func TestCSPC(t *testing.T) {
```
This is more like an integration test as it depends on >1 function, I'll refactor this later, such that an instance of `makePools` or `filterBDs` are called, this will also help expand the generation of other type of pool YAMLs or other similar resources such as JVP, SC(CStor & Jiva), etc.

vharsh commented on pkg/generate/testdata_test.go:
```diff
@@ -0,0 +1,197 @@
+      dataRaidGroupType: mirror
+  - dataRaidGroups:
+    - blockDevices:
+      # /dev/sda  1.0GiB
```
The size is displayed in a weird way now, I'll fix it sometime soon.

vharsh commented on pkg/generate/cspc.go:
```diff
@@ -0,0 +1,215 @@
+
+// filterCStorCompatible takes a list of BDs and gives out a list of BDs which can be used to provision a pool
+func filterCStorCompatible(bds *v1alpha1.BlockDeviceList) (*v1alpha1.BlockDeviceList, error) {
+	// TODO: Optionally reject sparse-disks depending on configs
+	var final []v1alpha1.BlockDevice
+	for _, bd := range bds.Items {
+		// an eligible blockdevice is in active+unclaimed state and lacks a file-system
+		if bd.Status.State == v1alpha1.BlockDeviceActive &&
+			bd.Status.ClaimState == v1alpha1.BlockDeviceUnclaimed &&
+			bd.Spec.FileSystem.Type == "" &&
+			// BD's capacity >=64 MiB
+			bd.Spec.Capacity.Storage >= 67110000 {
+			final = append(final, bd)
+		}
+	}
```
[CSTOR] Is this criteria sufficient or it should check the NDM config for `auto` & `manual` modes so that it can reject sparse files or something else?

mittachaitu replied:
Yeah, Good to reject sparse-based blockdevices(Good Point :+1: ),

**Just a optional suggestion**

We can replace the above snippet with something like below(it can be improved further):
```go
type poolBDFilter func(v1alpha1.BlockDevice) bool
var (
poolEligibleBDFilters := []poolBDFilter{
  isBlockDeviceActive,
  isBlockDeviceUnclaimed,
  isBlockDeviceUnFormatted,
  isNonSparseBlockDevice,
  isBlockDeviceHasMinCap(cap uint64) func(v1alpha1.BlockDevice) bool {
      return func(bd v1alpha1.BlockDevice) bool {
           return bd.Spec.Capacity.Storage >= minCapacity
      }
  }
}
)

func isBlockDeviceEligibleForPoolCreation(bd v1alpha1.BlockDevice, filters []poolBDFilter) bool {
             for i:=0; i<len(filters); i++ {
                   if !filters[i](bd) {
                       return false
                  }
             }
             return true
}

if isBlockDeviceEligibleForPoolCreation(bd, poolEligibleBDFilters) {
   final = append(final, bd)
}
```

vharsh commented on pkg/generate/testdata_test.go:
```diff
@@ -0,0 +1,197 @@
+			DataRaidGroups: []cstorv1.RaidGroup{{CStorPoolInstanceBlockDevices: []cstorv1.CStorPoolInstanceBlockDevice{{
+				BlockDeviceName: "bd1-n3"}, {BlockDeviceName: "bd2-n3"}}}}, PoolConfig: cstorv1.PoolConfig{DataRaidGroupType: string(cstorv1.PoolMirrored)}}}}}
+
+var mirrorCSPCstr = `apiVersion: cstor.openebs.io/v1
```
[CSTOR] Does this `mirror` configuration look good in case I've got two disks of 1GiB on each of 3 worker nodes with hostnames `, what more detail should I add in comments.
Some good suggestions can be

1. Actual size available to applications depending upon pool type.
2. PoolType features, when should one go for mirror, stripe, raidz, raidz2, etc.

vharsh commented on pkg/generate/testdata_test.go:
```diff
@@ -0,0 +1,197 @@
+    poolConfig:
+      dataRaidGroupType: mirror
+  - dataRaidGroups:
+    - blockDevices:
```
[CSTOR] The number of devices in a CSPIbd should be a multiple of two and there can be any number of `dataRaidGroups` 

Took this suggestion from [webhook validation](https://github.com/openebs/cstor-operators/blob/develop/pkg/webhook/cspc.go#L302), thanks @mittachaitu for pointing to the actual codebase.

Commit:
Fix stripe pool creation with multiple nodes & BDs

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

Commit:
Reduce LoC as stripe & mirror pool organization are similar

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

vharsh commented on cmd/generate/generate.go:
```diff
@@ -0,0 +1,59 @@
+	_ = cmd.MarkPersistentFlagRequired("nodes")
+	cmd.PersistentFlags().IntVar(&devices, "number-of-devices", 1, "number of devices per node")
+	cmd.PersistentFlags().StringVarP(&raidType, "raidtype", "", "stripe",
+		"allowed RAID configuration such as, stripe, mirror, raid, raidz2")
```
These are the exact poolType phrases/terms in flags & everywhere else.

vharsh commented on go.mod:
```diff
@@ -13,6 +13,7 @@ require (
 	github.com/pkg/errors v0.9.1
 	github.com/spf13/cobra v1.1.1
 	github.com/spf13/viper v1.7.0
+	github.com/stretchr/testify v1.6.1
```
for nested structures, JSONs, YAMLs, this seems to be the best library possible; other similar-ish lightweight libraries are [is](https://pkg.go.dev/github.com/matryer/is#section-readme) and a cool library which we might like better [gocmp](https://github.com/google/go-cmp/).

# Reasons for preferring go-cmp for comparing structs

1. [Cleaner one line-ish diffs of structs](https://blog.artwolf.in/a?ID=2e0cd2a0-3dfd-4dbc-a87d-c40647552b9a)
2. [Mentioned by Dave Cheney in testing blogs](https://dave.cheney.net/tag/testing)

mittachaitu commented on cmd/generate/generate.go:
```diff
@@ -0,0 +1,59 @@
+	return cmd
+}
+
+func NewCmdGenerateStorage() *cobra.Command {
```
nit: Good to have a function name more specific to CSPC/cStor pool instead of a generic name(in the future we can accommodate mayastor pools as well)
```suggestion
func NewCmdGenerateCstorStoragePool() *cobra.Command {
```

vharsh replied:
I'll make this change as we aren't sure if mayastor functionalities will be included in this CLI.

vharsh replied:
Changed it to `NewCmdGenerateCstorStoragePoolCluster` to avoid naming convention with earlier name of pool

mittachaitu commented on cmd/generate/generate.go:
```diff
@@ -0,0 +1,59 @@
+	var devices int
+	cmd := &cobra.Command{
+		Use:   "cspc",
+		Short: "Displays status information about Volume(s)",
```
Generates CSPC resources YAML/configuration which can be used to provision cStor storagepools

mittachaitu commented on cmd/generate/generate.go:
```diff
@@ -0,0 +1,59 @@
+	cmd.PersistentFlags().StringVarP(&nodes, "nodes", "", "",
+		"comma separated set of nodes for pool creation --nodes=node1,node2,node3,node4")
+	_ = cmd.MarkPersistentFlagRequired("nodes")
+	cmd.PersistentFlags().IntVar(&devices, "number-of-devices", 1, "number of devices per node")
```
Can we default it to no.of devices based on raidType? It will be helpful for users not to go through documentation

vharsh replied:
I'll see if a function taking `cmd` as an argument and returning the default minimum number of disks can be returned, good thought.

mittachaitu commented on cmd/generate/generate.go:
```diff
@@ -0,0 +1,59 @@
+			node, _ := cmd.Flags().GetString("nodes")
+			devs, _ := cmd.Flags().GetInt("number-of-devices")
+			raid, _ := cmd.Flags().GetString("raidtype")
+			Nodes := strings.Split(node, ",")
```
nit:
```suggestion
			nodeNameList := strings.Split(node, ",")
```

mittachaitu commented on pkg/generate/cspc.go:
```diff
@@ -0,0 +1,214 @@
+}
+
+// Pool calls the generate routine for different cas-types
+func Pool(nodes []string, devs int, raid string) error {
```
nit: Good to have function name related to CSPC/cStor pool
```suggestion
func CStorPool(nodes []string, devs int, raid string) error {
```

vharsh replied:
I have this function to call different generator functions for `cstor`, but it seems pool concept as such isn't there on non-cstor nodes wherein writing Pool YAMLs is complex.

vharsh replied:
I've renamed the functions to `CSPC` and `cspc`, they aren't good names, but having a function create a K8s-client object helps me better unit test it.

Background: The CSPC function is a dumb one, it does two things
1. Create a client object
2. Call the `cspc` which does interesting stuff and returns the `err` which is printed to the console.

Benefits: I can easily write the tests for business logic( which is in the `cspc(....)` and `CSPC()` has stuff which don't require tests). This way the `cmd` package has no k8s dependency and `cmd` package tests can also be written as an overkill someday, moreover, someone can import `openebsctl` as a dependency and call some of it's function(won't happen anyday I suppose).

mittachaitu commented on pkg/generate/cspc.go:
```diff
@@ -0,0 +1,214 @@
+func isPoolTypeValid(raid string) bool {
+	if raid == "stripe" || raid == "mirror" || raid == "raidz" || raid == "raidz2" {
+		return true
+	} else {
+		return false
+	}
```
```suggestion
	}
	return false
```

vharsh replied:
I wonder why golangci-lint doesn't raise such warnings :(

mittachaitu commented on pkg/generate/cspc.go:
```diff
@@ -0,0 +1,214 @@
+	// assume CSTOR's OPENEBS_NAMESPACE has all the relevant blockdevices
+	c.Ns = cstorNS
+	if err != nil {
+		return nil, "", fmt.Errorf("cannot find an active cstor installation")
```
```suggestion
		return nil, "", fmt.Errorf("unable to determine the cStor namespace error: %v", err)
```
We are swallowing the original error in [c.GetOpenEBSNamespace](https://github.com/openebs/openebsctl/blob/9b279df44ee1c8c9d69950b08015690c8a0944b2/pkg/client/k8s.go#L188) good to return different error based on err variable... Something like below:


```go
func (k K8sClient) GetOpenEBSNamespace(casType string) (string, error) {
	pods, err := k.K8sCS.CoreV1().Pods("").List(context.TODO(), metav1.ListOptions{FieldSelector: "status.phase=Running", LabelSelector: fmt.Sprintf("openebs.io/component-name=%s", util.CasTypeAndComponentNameMap[strings.ToLower(casType)])})
	if err != nil {
		return "", errors.Wrapf(err, "unable to determine openebs namespace")
	}
        if len(pods.Items) == 0 {
                return "", errors.New("unable to find cStor related artifacts")
        }
	return pods.Items[0].Namespace, nil
}
```

vharsh replied:
given the context is an empty one, i.e. it'll get stuck sort of forever and not get cancelled, what other `err` can come when `len(pod.Items)` is non zero?

vharsh replied:
Ah, got it, even when the pods won't exits, the `err` will likely be nil, I'll make the necessary changes.

mittachaitu commented on pkg/client/k8s.go:
```diff
@@ -493,3 +481,44 @@ func (k K8sClient) GetBatchJobs() (*batchV1.JobList, error) {
+	if len(nodes) == 0 {
+		return n, err
+	}
+	// 2. Put them in a map[string]*corev1.Node
```
```suggestion
	// 2. Put them in a map[string]corev1.Node
```

vharsh replied:
I'll put them in a map of string -> pointer to node instead, I wish there was an easy way to do this without resorting to unstructured objects(the java dev inside me misses generics).

mittachaitu commented on pkg/generate/cspc.go:
```diff
@@ -0,0 +1,214 @@
+	// 1. Validate nodes & poolType, fetch disks
+	nodeList, err := c.GetNodes(nodes, "", "")
+	if err != nil {
+		return nil, "", fmt.Errorf("(server error) unable to fetch nodes %s", err)
```
```suggestion
		return nil, "", fmt.Errorf("(server error) unable to fetch node information %s", err)
```

mittachaitu commented on pkg/generate/cspc.go:
```diff
@@ -0,0 +1,214 @@
+		return nil, "", fmt.Errorf("not all worker nodes are available for provisioning a CSPC")
+	}
+	// 2. Fetch BD's from the eligible/valid nodes
+	bds, err := c.GetBDs(nil, "kubernetes.io/hostname in ("+strings.Join(nodes, ",")+")")
```
This will not work in the case of EKS based clusters since in EKS setup nodeName is different from the `kubernetes.io/hostname` value. So better to get the hostNames for given nodeNames and then query for bds

vharsh replied:
Okay, that shouldn't be <10 lines of code, so I'll do the following for every-case, as it might not be straightforward to figure out if a cluster has this quirk. Found relevant bug @ `https://github.com/kubernetes/kubernetes/issues/54482`

1. Get all nodes by the node-name and then find the hostname and put them in a map[nodeName]HostName.
2. Assuming, `kubectl get nodes` gets me the `metadata.name` which is the nodeName and `metadata.Labels['kubernetes.io/hostname'] will get me the hostName, which I can map and then put them in a comma separated string

Or best, I don't even need a map 🤷🏼

mittachaitu commented on pkg/generate/cspc.go:
```diff
@@ -0,0 +1,214 @@
+	// 2. Fetch BD's from the eligible/valid nodes
+	bds, err := c.GetBDs(nil, "kubernetes.io/hostname in ("+strings.Join(nodes, ",")+")")
+	if err != nil || len(bds.Items) == 0 {
+		return nil, "", fmt.Errorf("no blockdevices found in %s nodes", nodes)
```
Let's also log about the error and do we need to change the format specifier to %v since nodes is a string slice

vharsh replied:
this error will get logged eventually as this line of errors aren't swallowed by the line of callers

mittachaitu commented on pkg/generate/cspc.go:
```diff
@@ -0,0 +1,214 @@
+		if nDevices%2 != 0 {
+			return nil, fmt.Errorf("mirrored pool requires multiples of two block device")
+		}
+		for node, bds := range bd {
```
```suggestion
		for hostName, bds := range bd {
```

mittachaitu commented on pkg/generate/cspc.go:
```diff
@@ -0,0 +1,214 @@
+		for node, bds := range bd {
+			var raids []cstorv1.CStorPoolInstanceBlockDevice
+			// add all BDs to a CSPCs CSPI spec
+			for d := 0; d < nDevices; d++ {
+				raids = append(raids, cstorv1.CStorPoolInstanceBlockDevice{BlockDeviceName: bds[d].Name})
+			}
```
I have little concern over here... Let's say that the user is given to pick 4(nDevices) devices from a node to form a pool then I will suggest forming (nDevices/2) raidgroups instead of accumulating on same raidGroup

```go
   // Ignore this check if it already made in caller functions
   // Check if required bd's exist for given node
   if len(bds) < nDevices {
      return errors.New("required no.of(%d) blockdevices doesn't exist on node %s", nDevices, hostName)
   }
   bdIndex := 0
   var rgs []cstorv1.RaidGroup
  for rgIndex :=0; rgIndex<(nDevices/2); rgIndex++ {
          var raids []cstorv1.CStorPoolInstanceBlockDevice
          for d:=0; d<defaultMirrorCount; d++ {
                 raids = append(raids, cstorv1.CStorPoolInstanceBlockDevice{BlockDeviceName: bds[bdIndex].Name})
                 bdIndex++
          }
          rgs = append(rgs, raids)
  }
```
WDYT on above part?


Basically, with the current approach if the user configures 4 devices of size 10G then the size of the pool will be 10G(n redundancy will exist at pool level)
```sh
sparse-01d1417508c6ee62a915bfd0c341c78f        centos-worker-1   /var/openebs/sparse/2-ndm-sparse.img                10737418240      Unclaimed    Active     3h20m
sparse-07a7fbcd6a14cd09e21a539734ca4772        centos-worker-1   /var/openebs/sparse/4-ndm-sparse.img                10737418240      Unclaimed    Active     40s
sparse-5ac9c0a5b87adeecba382f3ec1c093d5        centos-worker-1   /var/openebs/sparse/1-ndm-sparse.img                10737418240      Unclaimed    Active     3h20m
sparse-9903f755f31c265bb190df228f2d3221        centos-worker-1   /var/openebs/sparse/5-ndm-sparse.img                10737418240      Unclaimed    Active     40s
sparse-f62a0627a72f86deba09e73f8f2dc809        centos-worker-1   /var/openebs/sparse/3-ndm-sparse.img                10737418240      Unclaimed    Active     3h20m

## Creating mirror pool with 4 blockdevices
root@cstor-disk-cspc-v5jb-7f7d498b47-t8w58:/# zpool create pool1 mirror /var/openebs/sparse/2-ndm-sparse.img  /var/openebs/sparse/4-ndm-sparse.img /var/openebs/sparse/1-ndm-sparse.img /var/openebs/sparse/5-ndm-sparse.img

### size of the pool is 10G
root@cstor-disk-cspc-v5jb-7f7d498b47-t8w58:/# zfs list
NAME                                         USED  AVAIL  REFER  MOUNTPOINT
pool1                                         53K  9.63G   512B  /pool1

### Pool vdev will looks like
root@cstor-disk-cspc-v5jb-7f7d498b47-t8w58:/# zpool status -v
  pool: pool1
 state: ONLINE
  scan: none requested
config:

	NAME                                      STATE     READ WRITE CKSUM
	pool1                                     ONLINE       0     0     0
	  mirror-0                                ONLINE       0     0     0
	    /var/openebs/sparse/2-ndm-sparse.img  ONLINE       0     0     0
	    /var/openebs/sparse/4-ndm-sparse.img  ONLINE       0     0     0
	    /var/openebs/sparse/1-ndm-sparse.img  ONLINE       0     0     0
	    /var/openebs/sparse/5-ndm-sparse.img  ONLINE       0     0     0
```

With above suggestion it will be create pool of 20G with same configuration but redundancy of data will be only 1 copy than 3 copies
```sh
## Creating stripe pool on top of mirror pool
root@cstor-disk-cspc-v5jb-7f7d498b47-t8w58:/# zpool create pool1 mirror /var/openebs/sparse/2-ndm-sparse.img  /var/openebs/sparse/4-ndm-sparse.img mirror /var/openebs/sparse/1-ndm-sparse.img /var/openebs/sparse/5-ndm-sparse.img

## size of pool is 20G
root@cstor-disk-cspc-v5jb-7f7d498b47-t8w58:/# zfs list
NAME                                         USED  AVAIL  REFER  MOUNTPOINT
pool1                                         50K  19.3G   512B  /pool1

## pool vdev will looks like below
root@cstor-disk-cspc-v5jb-7f7d498b47-t8w58:/# zpool status
  pool: pool1
 state: ONLINE
  scan: none requested
config:

	NAME                                      STATE     READ WRITE CKSUM
	pool1                                     ONLINE       0     0     0
	  mirror-0                                ONLINE       0     0     0
	    /var/openebs/sparse/2-ndm-sparse.img  ONLINE       0     0     0
	    /var/openebs/sparse/4-ndm-sparse.img  ONLINE       0     0     0
	  mirror-1                                ONLINE       0     0     0
	    /var/openebs/sparse/1-ndm-sparse.img  ONLINE       0     0     0
	    /var/openebs/sparse/5-ndm-sparse.img  ONLINE       0     0     0

errors: No known data errors
```

vharsh replied:
Okay, so you mean, I should cap the number of BDs in a RaidGroup to 2? If a user gives 6 `nDevices`, 3 RaidGroups should be formed in each pool instance spec?

mittachaitu replied:
>  Okay, so you mean, I should cap the number of BDs in a RaidGroup to 2? If a user gives 6 nDevices, 3 RaidGroups should be formed in each pool instance spec?

Yes, this way cli will offer best to user out of given things

vharsh replied:
Got, it will do the same as suggested.

vharsh replied:
[Update] I've done most of the work, I'm done stuck between this nodename & hostname quirk. Usually peeps would talk in terms of nodeName(because that's what they see in `kubectl get nodes`), so it'd make sense I show errors with nodeNames. Aah, this can get tricky, I need a map of `nodeName->hostName` or vice-versa. 😞

vharsh replied:
Wrote similar logic, since I know that the number of BDs in a RaidGroup will be fixed to `2`, i.e. default one, I'm just adding two BDs into the RaidGroup slice at a time.

vharsh replied:
implemented similar logic to generate, RaidZ & RaidZ2 pools, I'll now have to do something to fix the capacity of the BDs which get selected in the pool assuming two cases
1. I have an infinite supply of BDs of different sizes
2. I have a finite supply of BDs and I should backtrack and re-select the most optimal ones which also minimize disk wastage

mittachaitu reviewed (commented):
Provided few comments/suggestions... PTAL

Commit:
Revert "Reduce LoC as stripe & mirror pool organization are similar"

This reverts commit 9efca730d90d1922eba08e051ca05acddd66ecda.

Why?

- RaidGroup configuration for `mirror` type pools aren't similar to
  `stripe` for the best possible redundancy, as rightly suggested by
  @mittachaitu

Commit:
Auto-select default number of devs for raidtype

* Update comments, rename functions

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

Commit:
Use hostname labels from nodes over nodeNames

* rename functions & variables, update comments

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

Commit:
Enhance error-handling from NodeName PoV

* User usually talks in terms of NodeName and not the node's hostname
* Iterate over a list of hostNames over a map[hostName][]BDs(which are
  unordered and result in flaky unit-tests)

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

Commit:
(mirror-pool) Keep two BDs in a RaidGroup.CSPIBD

if four BDs are assigned per node, the CSPC will have two RaidGroups
each having two BDs

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

vharsh commented on pkg/generate/testdata_test.go:
```diff
@@ -0,0 +1,301 @@
+			DataRaidGroups: []cstorv1.RaidGroup{{CStorPoolInstanceBlockDevices: []cstorv1.CStorPoolInstanceBlockDevice{{
+				BlockDeviceName: "bd1-n3"}, {BlockDeviceName: "bd2-n3"}}}}, PoolConfig: cstorv1.PoolConfig{DataRaidGroupType: string(cstorv1.PoolMirrored)}}}}}
+
+var mirrorCSPCFourBDs = cstorv1.CStorPoolCluster{
```
@mittachaitu Does this mirror CSPC on node1 & node2 look good where-in each RaidGroup has 2 blockdevices?

For six BDs, there'd be 3 RaidGroups per instance/Node, right?

mittachaitu replied:
Yes

mittachaitu commented on pkg/generate/cspc.go:
```diff
@@ -0,0 +1,243 @@
+	// 2. Fetch BD's from the eligible/valid nodes by hostname labels
+	bds, err := c.GetBDs(nil, "kubernetes.io/hostname in ("+strings.Join(hostnames, ",")+")")
+	if err != nil || len(bds.Items) == 0 {
+		return nil, "", fmt.Errorf("no blockdevices found in nodes with %v hostnames", hostnames)
```
Good to log error also

vharsh replied:
Returned errors are printed, the chain is a bit complex to the point it's not easy to figure these things out(which is not a good practice IMO) 😰 .

cspc.go:83 -> cspc.go:49 -> cmd/generate/generate.go:54 -> `util.CheckError is called with util.Fatal` and then this error is printed to stderr from pkg/util/util.go:74

mittachaitu reviewed (approved):
LGTM

mittachaitu reviewed (commented):
We missed picking same size of blockdevices and filtering sparse blockdevices instead of combining them with non-sparse blockdevices

vharsh commented:
> We missed picking same size of blockdevices and filtering sparse blockdevices instead of combining them with non-sparse blockdevices

@mittachaitu I'm planning to handle those things in upcoming PR as this PR has become a bit big
